### PR TITLE
Generate: consistently handle special tokens as tensors

### PR DESCRIPTION
**Pull request by gante**
# What does this PR do?

To enable `torch.compile` with `generate`, some special token-related operations have to be rewritten into torch operations. That requires special tokens to be tensors instead of integers or a list of integers. (See https://github.com/huggingface/transformers/pull/29374 for a working prototype)

This PR reworks special token usage in `generate` to consistently treat them as a tensor, as opposed to e.g. keeping track of `eos_token_id` in integer and in tensor form.

👉 Review suggestion: start by reading `_prepare_special_tokens` and how it fits in `generate`.

______________________________________________________________________
Requirements before merging this PR:
- [ ] merge https://github.com/huggingface/transformers/pull/29718 [Fixes doctests]

Tests ran locally:
- [ ] logits processors doctests (`pytest --doctest-modules src/transformers/generation/logits_process.py -vv`), needs requirement to be merged first
- [x] generate doctests (`pytest --doctest-modules src/transformers/generation/utils.py -vv`)
- [x] generate integration tests (`RUN_SLOW=1 py.test tests/generation/test_utils.py -vv`)
- [x] cache integration tests (`RUN_SLOW=1 py.test tests/test_cache_utils.py -vv`) -- same failures as in `main`
- [x] llama slow tests (`RUN_SLOW=1 py.test tests/models/llama/test_modeling_llama.py -vv`)
- [x] whisper slow tests (`RUN_SLOW=1 py.test tests/models/whisper/test_modeling_whisper.py -vv`) -- same failures as in `main`

### CODE REVIEW

File: src/transformers/generation/utils.py [hunk 1]
```diff
@@ -567,24 +570,6 @@ def _prepare_decoder_input_ids_for_generation(
 
         return decoder_input_ids, model_kwargs
 
-    def _get_decoder_start_token_id(
```
Comment (gante) [inline src/transformers/generation/utils.py:570]: The logic of this function is now within `_prepare_special_tokens`, which preprocesses all special tokens

File: src/transformers/generation/utils.py [hunk 2]
```diff
@@ -1221,6 +1208,55 @@ def _prepare_generation_config(
 
         return generation_config, model_kwargs
 
+    def _prepare_special_tokens(
```
Comment (gante) [inline src/transformers/generation/utils.py:1211]: ALL preprocessing logic for the special tokens now resides in this function 🧹

File: src/transformers/generation/utils.py [hunk 3]
```diff
@@ -1221,6 +1208,55 @@ def _prepare_generation_config(
...
+        self, generation_config: GenerationConfig, kwargs_has_attention_mask: Optional[bool] = None
```
Comment (gante) [inline src/transformers/generation/utils.py:1212]: `kwargs_has_attention_mask` is an optional argument so we can use this function in tests, to prepare special tokens.

File: src/transformers/generation/utils.py [hunk 4]
```diff
@@ -1914,12 +1934,17 @@ def _contrastive_search(
...
         logits_warper = logits_warper if logits_warper is not None else LogitsProcessorList()
         stopping_criteria = stopping_criteria if stopping_criteria is not None else StoppingCriteriaList()
-        pad_token_id = pad_token_id if pad_token_id is not None else self.generation_config.pad_token_id
-        eos_token_id = eos_token_id if eos_token_id is not None else self.generation_config.eos_token_id
         sequential = sequential if sequential is not None else self.generation_config.low_memory
-        if isinstance(eos_token_id, int):
-            eos_token_id = [eos_token_id]
-        eos_token_id_tensor = torch.tensor(eos_token_id).to(input_ids.device) if eos_token_id is not None else None
+
+        if not isinstance(pad_token_id, torch.Tensor):
```
Comment (gante) [inline src/transformers/generation/utils.py:1939]: The decoding functions are backward compatible (for now), and we can still pass `int`/`list(int)` as special tokens

File: src/transformers/models/musicgen/modeling_musicgen.py [hunk 5]
```diff
@@ -2170,6 +2170,24 @@ def _maybe_initialize_input_ids_for_generation(
                 break
         return torch.ones((batch_size, 1), dtype=torch.long, device=self.device) * bos_token_id
 
+    def _get_decoder_start_token_id(
```
Comment (gante) [inline src/transformers/models/musicgen/modeling_musicgen.py:2173]: Musicgen (and its melody variant) have their own custom `generate`, relying on this method. 

I've intentionally not updated this custom `generate`, to pressure us into moving towards a single generate function.

File: src/transformers/generation/utils.py [hunk 6]
```diff
@@ -454,20 +454,31 @@ def _maybe_initialize_input_ids_for_generation(
     def _prepare_attention_mask_for_generation(
         self,
         inputs: torch.Tensor,
-        pad_token_id: Optional[int],
-        eos_token_id: Optional[Union[int, List[int]]],
+        pad_token_id: Optional[torch.Tensor],
+        eos_token_id: Optional[torch.Tensor],
     ) -> torch.LongTensor:
+        # No information for attention mask inference -> return default attention mask
```
Comment (gante) [inline src/transformers/generation/utils.py:460]: The logic rewritten in functions like this is `torch.compile(..., fullgraph=True)` compatible 😉

File: src/transformers/generation/logits_process.py [hunk 7]
```diff
@@ -217,8 +225,8 @@ def __call__(self, input_ids: torch.LongTensor, scores: torch.FloatTensor) -> to
         new_tokens_length = input_ids.shape[-1] - self.prompt_length_to_skip
         scores_processed = scores.clone()
         vocab_tensor = torch.arange(scores.shape[-1], device=scores.device)
-        eos_token_id = torch.tensor(self.eos_token_id, device=scores.device)
```
Comment (zucchini-nlp) [inline src/transformers/generation/logits_process.py:220]: We have same conversion to tensor during `call` in suppress token processors, maybe can init them as tensors in `init` also.

**Review by zucchini-nlp — APPROVED** [review]
Thanks for working on this 😄

File: src/transformers/generation/utils.py [hunk 8]
```diff
@@ -3104,7 +3145,10 @@ def _beam_search(
             next_token_scores = next_token_scores.view(batch_size, num_beams * vocab_size)
 
             # Sample 1 + len(eos_token_id) next tokens for each beam so we have at least 1 non eos token per beam.
-            n_eos_tokens = len(eos_token_id) if eos_token_id else 0
+            if isinstance(eos_token_id, torch.Tensor):
+                n_eos_tokens = eos_token_id.shape[0]
+            else:
+                n_eos_tokens = len(eos_token_id) if eos_token_id else 0
             next_token_scores, next_tokens = torch.topk(
```
[the remark this answers is not in the document]
Comment (gante) [reply src/transformers/generation/utils.py:3152]: Good catch!

**Comment by gante** [comment]
let's merge https://github.com/huggingface/transformers/pull/29956 first, so the diff here becomes much smaller (the EOS-as-stopping-criteria made the diff more elaborate)
